src/Storages/StorageMergeTree.cpp

```diff
@@ -1463,8 +1463,11 @@ UInt64 StorageMergeTree::getCurrentMutationVersion(
 
 size_t StorageMergeTree::clearOldMutations(bool truncate)
 {
-    size_t finished_mutations_to_keep = truncate ? 0 : getSettings()->finished_mutations_to_keep;
+    size_t finished_mutations_to_keep = getSettings()->finished_mutations_to_keep;
+    if (!finished_mutations_to_keep)
+        return 0;
 
+    finished_mutations_to_keep = truncate ? 0 : finished_mutations_to_keep;
```

Comment:
As I see we call this function with `truncate = true` when
1. the table is truncated
2. a part or partition is dropped

In case 1 I think it make sense to really truncate the old mutations, however it is not clear from the docs what is the expected behavior.

In case 2 it is questionable. To be honest I don't see a good reason why to clean the old mutations when a part or partition is droopped.

Based on these paprtial information, I have to ask: what do you think about enforcing the truncate flag regardless the value of `finished_mutations_to_keep`, but only call this function with `truncate = true` when we truncate a table?

Reply:
Indeed, make total sense. Rebased.

_P.S. case 1 I broke unintentionally_